File: android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorage.java
package com.oblador.keychain.cipherStorage;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.oblador.keychain.SecurityLevel;
import com.oblador.keychain.decryptionHandler.DecryptionResultHandler;
import com.oblador.keychain.exceptions.CryptoFailedException;
import com.oblador.keychain.exceptions.KeyStoreAccessException;

import java.security.Key;
import java.util.Set;

@SuppressWarnings({"unused", "WeakerAccess"})
public interface CipherStorage {
  //region Helper classes

  /** basis for storing credentials in different data type formats. */
  abstract class CipherResult<T> {
    public final T username;
    public final T password;

    public CipherResult(final T username, final T password) {
      this.username = username;
      this.password = password;
    }
  }

  /** Credentials in bytes array, often a result of encryption. */
  class EncryptionResult extends CipherResult<byte[]> {
    /** Name of used for encryption cipher storage. */
    public final String cipherName;

    /** Main constructor. */
    public EncryptionResult(final byte[] username, final byte[] password, final String cipherName) {
      super(username, password);
      this.cipherName = cipherName;
    }

    /** Helper constructor. Simplifies cipher name extraction. */
    public EncryptionResult(final byte[] username, final byte[] password, @NonNull final CipherStorage cipherStorage) {
      this(username, password, cipherStorage.getCipherStorageName());
    }
  }

  /** Credentials in string's, often a result of decryption. */
  class DecryptionResult extends CipherResult<String> {
    private final SecurityLevel securityLevel;

    public DecryptionResult(final String username, final String password) {
      this(username, password, SecurityLevel.ANY);
    }

    public DecryptionResult(final String username, final String password, final SecurityLevel level) {
      super(username, password);
      securityLevel = level;
    }

    public SecurityLevel getSecurityLevel() {
      return securityLevel;
    }
  }

  /** Ask access permission for decrypting credentials in provided context. */
  class DecryptionContext extends CipherResult<byte[]> {
    public final Key key;
    public final String keyAlias;

    public DecryptionContext(@NonNull final String keyAlias,
                             @NonNull final Key key,
                             @NonNull final byte[] password,
                             @NonNull final byte[] username) {
      super(username, password);
      this.keyAlias = keyAlias;
      this.key = key;
    }
  }

  //region API

  /** Encrypt credentials with provided key (by alias) and required security level. */
  @NonNull
  EncryptionResult encrypt(@NonNull final String alias,
                           @NonNull final String username,
                           @NonNull final String password,
                           @NonNull final SecurityLevel level)
    throws CryptoFailedException;

  /**
   * Decrypt credentials with provided key (by alias) and required security level.
   * In case of key stored in weaker security level than required will be raised exception.
   * That can happens during migration from one version of library to another.
   */
  @NonNull
  DecryptionResult decrypt(@NonNull final String alias,
                           @NonNull final byte[] username,
                           @NonNull final byte[] password,
                           @NonNull final SecurityLevel level)
    throws CryptoFailedException;

  /** Decrypt the credentials but redirect results of operation to handler. */
  void decrypt(@NonNull final DecryptionResultHandler handler,
               @NonNull final String alias,
               @NonNull final byte[] username,
               @NonNull final byte[] password,
               @NonNull final SecurityLevel level)
    throws CryptoFailedException;

  /** Remove key (by alias) from storage. */
  void removeKey(@NonNull final String alias) throws KeyStoreAccessException;

  /**
   * Return all keys present in this storage.
   * @return key aliases
   */
  Set<String> getAllKeys() throws KeyStoreAccessException;

  //endregion

  //region Configuration

  /** Storage name. */
  String getCipherStorageName();

  /** Minimal API level needed for using the storage. */
  int getMinSupportedApiLevel();

  /** Provided security level. */
  SecurityLevel securityLevel();

  /** True - based on secured hardware capabilities, otherwise False. */
  boolean supportsSecureHardware();

  /** True - based on biometric capabilities, otherwise false. */
  boolean isBiometrySupported();

  /**
   * The higher value means better capabilities.
   * Formula:
   * = 1000 * isBiometrySupported() +
   * 100 * isSecureHardware() +
   * minSupportedApiLevel()
   */
  int getCapabilityLevel();

  /** Get default name for alias/service. */
  String getDefaultAliasServiceName();
  //endregion
}

File: android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorage.kt
package com.oblador.keychain.cipherStorage

import com.oblador.keychain.SecurityLevel
import com.oblador.keychain.resultHandler.ResultHandler
import com.oblador.keychain.exceptions.CryptoFailedException
import com.oblador.keychain.exceptions.KeyStoreAccessException

@Suppress("unused", "MemberVisibilityCanBePrivate")
interface CipherStorage {

  // region Helper classes

  /** Basis for storing credentials in different data type formats. */
  abstract class CipherResult<T>(val username: T, val password: T)

  /** Credentials in bytes array, often a result of encryption. */
  class EncryptionResult(
    username: ByteArray,
    password: ByteArray,
    /** Name of cipher storage used for encryption. */
    val cipherName: String
  ) : CipherResult<ByteArray>(username, password) {
    /** Helper constructor. Simplifies cipher name extraction. */
    constructor(
      username: ByteArray,
      password: ByteArray,
      cipherStorage: CipherStorage
    ) : this(username, password, cipherStorage.getCipherStorageName())
  }

  /** Credentials in strings, often a result of decryption. */
  class DecryptionResult(
    username: String,
    password: String,
    private val securityLevel: SecurityLevel = SecurityLevel.ANY
  ) : CipherResult<String>(username, password) {
    fun getSecurityLevel(): SecurityLevel = securityLevel
  }

  // endregion

  // region API

  /**
   * Encrypt credentials with provided key (by alias) and required security level.
   *
   * @throws CryptoFailedException If encryption fails.
   */

  @Throws(CryptoFailedException::class)
  fun encrypt(
    handler: ResultHandler,
    alias: String,
    username: String,
    password: String,
    level: SecurityLevel
  )


  /**
   * Decrypt the credentials but redirect results of operation to handler.
   *
   * @throws CryptoFailedException If decryption fails.
   */
  @Throws(CryptoFailedException::class)
  fun decrypt(
    handler: ResultHandler,
    alias: String,
    username: ByteArray,
    password: ByteArray,
    level: SecurityLevel
  )

  /** Remove key (by alias) from storage. */
  @Throws(KeyStoreAccessException::class)
  fun removeKey(alias: String)

  /**
   * Return all keys present in this storage.
   *
   * @return Set of key aliases.
   */
  @Throws(KeyStoreAccessException::class)
  fun getAllKeys(): Set<String>

  // endregion

  // region Configuration

  /** Storage name. */
  fun getCipherStorageName(): String

  /** Minimal API level needed for using the storage. */
  fun getMinSupportedApiLevel(): Int

  /** Provided security level. */
  fun securityLevel(): SecurityLevel

  /** True if auth is supported. */
  fun isAuthSupported(): Boolean

  /**
   * The higher value means better capabilities. Formula: = 1000 * isBiometrySupported() + 100 *
   * supportsSecureHardware() + minSupportedApiLevel
   */
  fun getCapabilityLevel(): Int

  /** Get default name for alias/service. */
  fun getDefaultAliasServiceName(): String

  // endregion
}
